queasars/job_shop_scheduling/scip_solver.py

```diff
@@ -0,0 +1,136 @@
+        self._machine_operations: dict[Machine, list[Operation]] = {}
+
+    def get_model(self) -> Model:
+        """
+        Creates and returns a pyscipopt SCIP optimization model. Subsequent calls return the cached model
+
+        :return: a pyscipopt SCIP optimization model
+        :rtype: Model
+        """
+        if not self._model_prepared:
+            self._model = Model()
+            self._prepare_variables()
+            self._prepare_constraints()
+            self._model_prepared = True
+        return self._model
+
+    def parse_solution(self, solution: Solution) -> JobShopSchedulingResult:
+        """
+        Parses a pyscipopt solution for this encoder's optimization model
+
+        :param solution: solution obtained by the SCIP solver
+        :type solution: Solution
+        :return: a JobShopSchedulingResult parsed from the solution
+        :rtype: JobShopSchedulingResult
+        """
+        job_schedules: dict[Job, tuple[ScheduledOperation, ...]] = {}
+        for job in self._jssp_instance.jobs:
+            scheduled_operations: list[ScheduledOperation] = []
+            for operation in job.operations:
+                start_time: int = int(solution[self._operation_start_variables[operation]])
+                scheduled_operations.append(ScheduledOperation(operation=operation, start=start_time))
+            job_schedules[job] = tuple(scheduled_operations)
+
+        return JobShopSchedulingResult(problem_instance=self._jssp_instance, schedule=job_schedules)
+
+    def _prepare_variables(self):
+        """
+        Creates the integer variables for the starting times of each operation and constrains them to be at least 0.
+        Also creates a integer variable to hold the makespan, which is constrained to be larger than the end times
+        of the last operations. Sets the optimization goal to minimize the makespan variable
+        """
+        # create auxiliary variable to hold the makespan as an optimization goal
+        self._optimization_var = self._model.addVar("optimization_var", vtype="INTEGER")
+        self._model.setObjective(self._optimization_var)
+        self._model.addCons(self._optimization_var >= 0)
+
+        for job in self._jssp_instance.jobs:
+            for operation in job.operations:
+                # Cache the operations per machine
+                # This is needed for the additional constraints
+                if operation.machine not in self._machine_operations:
+                    self._machine_operations[operation.machine] = []
+                self._machine_operations[operation.machine].append(operation)
+
+                # Create an integer variable for the start time of each operation
+                # Constrain it to be at least 0
+                var = self._model.addVar(operation.identifier, vtype="INTEGER")
+                self._operation_start_variables[operation] = var
+                self._model.addCons(var >= 0)
```

Comment:
Not sure if this makes a difference, but I think the `addVar` method offers this functionality directly. In fact, according to the online documentation the default lower bound is zero, so you should be able to just delete this line. (maybe verify this though)

Reply:
Ah i see. I removed these lower bound constraints and now directly set the lower bound with the lb parameter in the variable creation.